### ISSUE TITLE
ABI & SONAME bumps

### ISSUE THREAD

**Issue by paravoid** [kristapsdz/lowdown#121]
I've noticed that the SONAME has been bumped to liblowdown.so.3 in 1.0.2 (with 1c6fc7e20c04b4168bb53ed434cea1b5a6508e18) and to .2 with 1.0.1 (with 492fd851a4b34ba69405461c3875bb9ba67f02cb).

SONAME bumps can get tricky for distributions, as that implies a changed ABI, and by extension that every packaged software that has been linking with liblowdown needing to be recompiled to link with the new shared object. In Debian we call these "library transitions". By extension, Debian packages are named after the SONAME, in order to map these linker relationships into package manager relationships, and to allow coinstallability (and avoid DLL hell). So, for example, `liblowdown.so.1` is included in the package `liblowdown1`.

In practice we don't have any software packaged in Debian that's built against liblowdown right now, so a transition would be small and manageable. There may be software out there installed locally by users, however, and it could be annoying for them.

In practice, every ABI bump means a SONAME bump, and every SONAME bump, a package name bump. It can get a bit complicate, and that's why techniques like symbol versioning exist.

So I guess it boils down to a question:
*  Did the ABI indeed change between 1.0.0 and 1.0.2? Would software linked against 1.0.0 not work against 1.0.2? In other words, Is the SONAME bump warranted? If not, should we revert it?

...and a request:
* Could we avoid breaking the ABI if possible? That means adding new API calls but keeping the old ones, or adding fields at the end of structs etc.  Symbol versioning can also be employed as an alternative.

I'll hold off on the 1.0.2 upload to Debian until we resolve this, to avoid a 1->3->1 library transition.

Thanks as always :)

**Comment by paravoid** [comment]
Given it's been a while and I was impatient enough to see 1.0.2 in Debian ;) I ran a few tests. I tried linking lowdown with liblowdown.so (instead of the .a), and then checking whether the test suite with lowdown 1.0.0 + liblowdown 1.0.2 continues to work (it does). I then proceeded to upload to Debian 1.0.2 patched with a LIBVER of 1.

The only issue I saw in my review as potentially problematic is that in 1.0.2 there was a `flags` field added at the beginning of `struct rndr_normal_text`. I considered moving it at the end, but the (only) other field, `text`, is a struct too (`struct lowdown_buf`), so that seems problematic as well (e.g. if a field gets added to that struct. Correct me if I'm wrong, but my impression is that `struct rndr_normal_text` can be used by liblowdown users (and thus could be considered as part of the ABI) but it'd be unusual to do so (so more of an internal thing).

**Comment by paravoid** [comment]
Any thoughts here? And any plans for a new release? Thanks!

**Comment by kristapsdz** [comment]
Hi Faidon!  If I were you, I'd just patch to always have LIBVER be at 1 and leave managing the changes to consumers of the library directly, of which there are few.

I'm trying to follow semver (the upcoming will bump minor to "1" to "1.1.0" due to API changes) and proper versioning in LIBVER (which is the major version) (now set at "4").

I'd like to have the new version out by tonight or next weekend, after having AFL chew on it for a while.

**Comment by paravoid** [comment]
First of all, if possible I'd like to avoid deltas with upstream :) I think there's value in having an agreement of what the SONAME is and sticking to it. In theory this allows portability of binaries across different Linux distributions as well, not just within Debian. While in practice I'd guess most would recompile from source, I think it's best if we avoid having a liblowdown binary link against liblowdown.so.1 in Debian, but liblowdown.so.4 in Fedora.

Second, with regards to SemVer and using the example of 1.1.0... are you changing the ABI in an incompatible way? I saw some commits fly by that added new API functions (`char_subscript`) but did not remove any. That would qualify as a "backwards compatible" SemVer bump, so 1.1.x, not 2.x, right? These would not break existing binaries i.e. the ABI, hence no need for a LIBVER bump. Am I missing something?

In other words, perhaps a good approximation is that LIBVER should follow the SemVer Major version, i.e. stay at 1 for the time being. (The "approximation" part is that you could technically still change the API in backwards compatibly ways, but not the ABI, e.g. through symbol versioning).

For new functions (symbols), we have tooling in Debian to track each symbol separately, and, for example if a liblowdown user utilizes `char_subscript`, to generate dependencies such as `Depends: liblowdown1 (>= 1.1.0)`. See https://salsa.debian.org/debian/lowdown/-/blob/debian/sid/debian/liblowdown1.symbols for how this looks like on the packaging side.

**Comment by kristapsdz** [comment]
Thank you so much for looking into this!  I really appreciate your efforts.

I'm not adding or removing any visible API functions---it's the structure fields that have changed by way of addition.  Really, diffing `lowdown.h` will tell all.  Regardless, these are still API changes.  This is why I incremented the minor semver, not the major.  (The functions added aren't visible with the new `#pragma` changes.)

For now, I'm going to stick with a `LIBVER` of 1, and yes, tie `LIBVER` to the major number.  Thanks for all of this clarification.

**Comment by paravoid** [comment]
Thanks so much :)

**Comment by robryk** [comment]
Please note that any changes to layout of structs that are used in the external interface is an ABI change (with some tricky exceptions around adding fields to ends of structs that are never allocated by the client). For example, 9f3a5d707287c245ec65bd7fc2511612fa457f2e (which did happen before this discussion to be fair) did make such a change, because it changed the size of that union and thus (a) changed lowdown_opts size (b) shifted fields in it.
